// minispec/dist/cjs/engine/runtime.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var context_js_1 = require("../runnables/context.js");
var test_js_1 = require("../runnables/test.js");
var hook_js_1 = require("../runnables/hook.js");
var Runtime = /** @class */ (function () {
    function Runtime(engine, activeContext) {
        this.engine = engine;
        this.activeContext = activeContext;
        this.contexts = [];
        this.hasSomeFocus = false;
    }
    Runtime.prototype.discover = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _i, _a, context;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _i = 0, _a = this.contexts;
                        _b.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3 /*break*/, 4];
                        context = _a[_i];
                        return [4 /*yield*/, context.discover()];
                    case 2:
                        _b.sent();
                        _b.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    Runtime.prototype.execute = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _i, _a, context;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _i = 0, _a = this.contexts;
                        _b.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3 /*break*/, 4];
                        context = _a[_i];
                        return [4 /*yield*/, context.execute(this.hasSomeFocus)];
                    case 2:
                        _b.sent();
                        _b.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    Runtime.prototype.addContext = function (name, code, options) {
        var context = new context_js_1.default(name, code, {
            focused: (options === null || options === void 0 ? void 0 : options.focused) || false,
            skipped: (options === null || options === void 0 ? void 0 : options.skipped) || false,
            engine: this.engine,
        });
        this.hasSomeFocus = this.hasSomeFocus || (options === null || options === void 0 ? void 0 : options.focused) || false;
        if (this.activeContext.hasOne()) {
            this.activeContext.get().addChildContext(context);
            return;
        }
        this.contexts.push(context);
    };
    Runtime.prototype.addTest = function (name, code, options) {
        var test = new test_js_1.default(name, code, {
            engine: this.engine,
            skipped: (options === null || options === void 0 ? void 0 : options.skipped) || false,
            location: (options === null || options === void 0 ? void 0 : options.location) || undefined
        });
        this.activeContext.get().addTest(test, (options === null || options === void 0 ? void 0 : options.focused) || false);
        this.hasSomeFocus = this.hasSomeFocus || (options === null || options === void 0 ? void 0 : options.focused) || false;
    };
    Runtime.prototype.addBeforeAllHook = function (code) {
        var hook = new hook_js_1.default(code);
        this.activeContext.get().addBeforeAllHook(hook);
    };
    Runtime.prototype.addAfterAllHook = function (code) {
        var hook = new hook_js_1.default(code);
        this.activeContext.get().addAfterAllHook(hook);
    };
    Runtime.prototype.addBeforeEachHook = function (code) {
        var hook = new hook_js_1.default(code);
        this.activeContext.get().addBeforeEachHook(hook);
    };
    Runtime.prototype.addAfterEachHook = function (code) {
        var hook = new hook_js_1.default(code);
        this.activeContext.get().addAfterEachHook(hook);
    };
    return Runtime;
}());
exports.default = Runtime;
