// minispec/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.afterAll = exports.afterEach = exports.beforeEach = exports.beforeAll = exports.sit = exports.xit = exports.fit = exports.it = exports.scontext = exports.xcontext = exports.fcontext = exports.context = exports.sdescribe = exports.xdescribe = exports.fdescribe = exports.describe = void 0;
var describe_js_1 = require("./api/describe.js");
exports.describe = describe_js_1.default;
Object.defineProperty(exports, "fdescribe", { enumerable: true, get: function () { return describe_js_1.fdescribe; } });
Object.defineProperty(exports, "xdescribe", { enumerable: true, get: function () { return describe_js_1.xdescribe; } });
Object.defineProperty(exports, "sdescribe", { enumerable: true, get: function () { return describe_js_1.sdescribe; } });
var context_js_1 = require("./api/context.js");
exports.context = context_js_1.default;
Object.defineProperty(exports, "fcontext", { enumerable: true, get: function () { return context_js_1.fcontext; } });
Object.defineProperty(exports, "xcontext", { enumerable: true, get: function () { return context_js_1.xcontext; } });
Object.defineProperty(exports, "scontext", { enumerable: true, get: function () { return context_js_1.scontext; } });
var it_js_1 = require("./api/it.js");
exports.it = it_js_1.default;
Object.defineProperty(exports, "fit", { enumerable: true, get: function () { return it_js_1.fit; } });
Object.defineProperty(exports, "xit", { enumerable: true, get: function () { return it_js_1.xit; } });
Object.defineProperty(exports, "sit", { enumerable: true, get: function () { return it_js_1.sit; } });
var before_all_js_1 = require("./api/before_all.js");
exports.beforeAll = before_all_js_1.default;
var before_each_js_1 = require("./api/before_each.js");
exports.beforeEach = before_each_js_1.default;
var after_each_js_1 = require("./api/after_each.js");
exports.afterEach = after_each_js_1.default;
var after_all_js_1 = require("./api/after_all.js");
exports.afterAll = after_all_js_1.default;
var engine_js_1 = require("./engine/engine.js");
var MiniSpec = engine_js_1.default;
exports.default = MiniSpec;
